### PR TITLE
Dynamodb

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

**Comment by sdelamo** [comment]
I tested this approach with test resources. It can be easily simplify by porting it to test-resources and using the [Generic Testcontainers support](https://micronaut-projects.github.io/micronaut-test-resources/snapshot/guide/#modules-testcontainers)

**Comment by timyates** [comment]
Created an issue in starter to create a ksuid feature

https://github.com/micronaut-projects/micronaut-starter/issues/1308

**Comment by timyates** [comment]
I ignored the failing test in authentication, and raised an issue to investigate https://github.com/micronaut-projects/micronaut-guides/issues/998 I cannot currently see why it fails, and it obviously passed once as we have a cached passing version 🤔

**Comment by timyates** [comment]
> Created an issue in starter to create a ksuid feature
> 
> [micronaut-projects/micronaut-starter#1308](https://github.com/micronaut-projects/micronaut-starter/issues/1308)

Or if it's only meant to be here to build the guide, we need to remove it from the list of features in the create section of the doc

image

**Review by timyates — APPROVED** [review]
Just a question around ksuid, but that shouldn't block this and I've raised an issue if we want to make it a feature.

If we don't want to make it a feature, we need to change the CLI line to exclude ksuid from the features

**Comment by sdelamo** [comment]
> Just a question around ksuid, but that shouldn't block this and I've raised an issue if we want to make it a feature.
> 
> If we don't want to make it a feature, we need to change the CLI line to exclude ksuid from the features

I have changed the CLI line to set only `dynamodb` feature.

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+common:build-lang-arguments.adoc[]
+
+common:default-package.adoc[]
+
+diffLink:[]
+
+=== Dynamo DB Dependencies
+
+To use DynamoDB with Micronaut, you application should have the following dependencies:
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc]:
```suggestion
To use DynamoDB with the Micronaut framework, your application should have the following dependencies:
```

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+callout:requires-property[1]
+callout:configuration-properties-interface[2]
+
+=== Repository
+
+Create an interface to encapsulate `Book` persistence.
+
+source:BookRepository[]
+
+Create a singleton class to handle common operation with DynamoDB.
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc]:
```suggestion
Create a singleton class to handle common operations with DynamoDB.
```

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+==== Dev default environment
+
+Modify `Application` to use `dev` as a https://docs.micronaut.io/latest/guide/index.html#_default_environment[default environment].
+
+source:Application[]
+
+==== Dev specific configuration
+
+Create an environment specific configuration class for the `dev` environment which defines the host and port of the DynamoDB Local instance.
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc]:
```suggestion
Create an environment-specific configuration class for the `dev` environment that defines the host and port of the DynamoDB local instance.
```

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+resource:application-dev.yml[]
+
+==== Dev Bootstrap
+
+Create a `StartupEventListener` only loaded for the `dev` environment which creates the dynamodb table if it does not exists.
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc]:
```suggestion
Create a `StartupEventListener` that is loaded only for the `dev` environment that creates a dynamodb table if one does not already exist.
```

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+source:DevBootstrap[]
+
+callout:requires-property[1]
+callout:requires-env[2]
+callout:singleton[3]
+
+==== Pointing to DynamoDB Local
+
+Create a bean created listener which points the DynamoDB client to the URL of the Dynamodb local instance.
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc]:
```suggestion
Add a bean-created listener that points the DynamoDB client to the URL of the Dynamodb local instance.
```

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+----
+
+[source,json]
+----
+[{"id":"2BLCWltdt3gGgSw1qsomXIfXBiX","isbn":"1680502395","name":"Release It!"}]
+----
+
+=== Tests
+
+Create a `StartupEventListener` only loaded for the `test` environment which creates the dynamodb table if it does not exists.
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc:191]:
```suggestion
Create a `StartupEventListener` that is loaded only for the `test` environment that creates a dynamodb table if one does not already exist.
```

File: guides/micronaut-dynamodb/micronaut-dynamodb.adoc [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+test:TestBootstrap[]
+
+Create a test which verifies the CRUD functionality.
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/micronaut-dynamodb.adoc:195]:
```suggestion
Create a test that verifies CRUD functionality.
```

File: guides/micronaut-rest-assured/metadata.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "title": "Using Micronaut Test REST-Assured in a Micronaut application",
```
Comment (kroegerk) [inline guides/micronaut-rest-assured/metadata.json:2]:
```suggestion
  "title": "Using Micronaut Test REST-Assured in a Micronaut Application",
```

File: src/docs/common/callouts/callout-bean-created-event-listener.adoc [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Creating a `@Singleton` which implements `BeanCreatedEventListener` allows you to provide extra configuraration to
```
Comment (kroegerk) [inline src/docs/common/callouts/callout-bean-created-event-listener.adoc]:
```suggestion
Creating a `@Singleton` that implements `BeanCreatedEventListener` allows you to provide extra configuration to
```

File: src/docs/common/snippets/common-ksuid.adoc [hunk 10]
```diff
@@ -0,0 +1,8 @@
+=== Ksuid
+
+Add the following dependency, to generate KSUIDs (K-Sortable Globally Unique IDs)
```
Comment (kroegerk) [inline src/docs/common/snippets/common-ksuid.adoc]:
```suggestion
Add the following dependency to generate KSUIDs (K-Sortable Globally Unique IDs).
```

File: src/docs/common/snippets/common-run-dynamodb-with-docker.adoc [hunk 11]
```diff
@@ -0,0 +1,11 @@
+For local develoment, we can https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html[setup DynamoDB locally].
```
Comment (kroegerk) [inline src/docs/common/snippets/common-run-dynamodb-with-docker.adoc]:
```suggestion
For local development, we can https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html[set up DynamoDB locally].
```

File: src/docs/common/snippets/common-run-dynamodb-with-docker.adoc [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+An ease option is to run it via Docker.
```
Comment (kroegerk) [inline src/docs/common/snippets/common-run-dynamodb-with-docker.adoc]:
```suggestion
An easy option is to run it via Docker.
```

File: src/docs/common/snippets/common-run-dynamodb-with-docker.adoc [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+After installing Docker, execute the following command to run a DynamoDB Local container:
```
Comment (kroegerk) [inline src/docs/common/snippets/common-run-dynamodb-with-docker.adoc]:
```suggestion
After installing Docker, execute the following command to run a DynamoDB local container:
```

File: guides/micronaut-dynamodb/metadata.json [hunk 14]
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Using DynamoDB in a Micronaut Application",
+  "intro": "Learn how to use DynamoDB as your persistence solution in a Micronaut Application.",
```
Comment (kroegerk) [inline guides/micronaut-dynamodb/metadata.json:3]: I just noticed that "application" in the "intro" line needs to be in lowercase.

File: src/docs/common/snippets/common-ksuid.adoc [hunk 15]
```diff
@@ -0,0 +1,8 @@
+=== Ksuid
+
+Add the following dependency to generate KSUIDs (K-Sortable Globally Unique IDs).
+____
+KSUID is for K-Sortable Unique IDentifier. It's a way to generate globally unique IDs similar to RFC 4122 UUIDs, but contain a time component so they can be "roughly" sorted by time of creation. The remainder of the KSUID is randomly generated bytes.
```
Comment (kroegerk) [inline src/docs/common/snippets/common-ksuid.adoc:5]: KSUID is for K-Sortable Unique IDentifier. It's a way to generate globally unique IDs similar to RFC 4122 UUIDs that contain a time component, so they can be "roughly" sorted by time of creation. The remainder of the KSUID is made up of randomly generated bytes.

**Review by kroegerk — COMMENTED** [review]
A couple more little changes.

**Pull request merged**